### PR TITLE
[Lake] Support snapshot

### PR DESCRIPTION
**Pull request by TszKitLo40**
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

Support snapshot
## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

File: be/src/service/service_be/lake_service.cpp [hunk 1]
```diff
@@ -311,4 +311,69 @@ void LakeServiceImpl::get_tablet_stats(::google::protobuf::RpcController* contro
...
+        return;
+    }
+    if (!request->has_tablet_id()) {
+        cntl->SetFailed("missing tablet id");
+        return;
+    }
+
+    auto tablet = _env->lake_tablet_manager()->get_tablet(request->tablet_id());
+    if (!tablet.ok()) {
+        LOG(ERROR) << "Fail to get tablet " << request->tablet_id() << ": " << tablet.status();
```
Comment (tracymacding) [inline be/src/service/service_be/lake_service.cpp:335]: cntl->SetFailed(...)

File: be/src/service/service_be/lake_service.cpp [hunk 2]
```diff
@@ -311,4 +311,69 @@ void LakeServiceImpl::get_tablet_stats(::google::protobuf::RpcController* contro
...
```
Comment (tracymacding) [inline be/src/service/service_be/lake_service.cpp:370]: cntl->SetFailed(...)

File: be/src/storage/lake/gc.cpp [hunk 3]
```diff
@@ -34,12 +34,17 @@ Status metadata_gc(std::string_view root_location, TabletManager* tablet_mgr) {
...
+    std::vector<string> txn_logs;
+    std::unordered_map<int64_t, std::vector<int64_t>> locked_tablet_metadatas;
 
     auto iter_st = fs->iterate_dir(std::string(root_location), [&](std::string_view name) {
         if (is_tablet_metadata(name)) {
             auto [tablet_id, version] = parse_tablet_metadata_filename(name);
             tablet_metadatas[tablet_id].emplace_back(version);
         }
+        if (is_txn_log(name)) {
+            txn_logs.emplace_back(name);
```
Comment (tracymacding) [inline be/src/storage/lake/gc.cpp:46]: Dangerous: you can not delete txn_id file because it may be used by load task

Comment (TszKitLo40) [reply be/src/storage/lake/gc.cpp:46]: `txn_log` won't be deleted here. The `txn_log` with metadata lock will be deleted in the release snapshots phase.

File: gensrc/proto/lake_service.proto [hunk 4]
```diff
@@ -88,6 +88,28 @@ message TabletStatResponse {
     repeated TabletStat tablet_stats = 1;
 }
 
+message TabletMetadataLockRequest {
```
[the remark this answers is not in the document]
Comment (TszKitLo40) [reply gensrc/proto/lake_service.proto:91]: OK, I will also rename `ReleaseTabletMetadataLockRequest` to `UnlockTabletMetaRequest`.

**Comment by silverbullet233** [comment]
run SonarCloud Code Analysis

File: be/src/service/service_be/lake_service.cpp [hunk 5]
```diff
@@ -311,4 +311,71 @@ void LakeServiceImpl::get_tablet_stats(::google::protobuf::RpcController* contro
...
+    }
+}
+
+void LakeServiceImpl::unlock_tablet_metadata(::google::protobuf::RpcController* controller,
+                                             const ::starrocks::lake::UnlockTabletMetadataRequest* request,
+                                             ::starrocks::lake::UnlockTabletMetadataResponse* response,
+                                             ::google::protobuf::Closure* done) {
+    brpc::ClosureGuard guard(done);
+    auto cntl = static_cast<brpc::Controller*>(controller);
+    if (!request->has_version()) {
```
Comment (wyb) [inline be/src/service/service_be/lake_service.cpp:356]: `version` is useless.

File: be/src/storage/lake/gc.cpp [hunk 6]
```diff
@@ -34,12 +34,17 @@ Status metadata_gc(std::string_view root_location, TabletManager* tablet_mgr) {
     }
 
     std::unordered_map<int64_t, std::vector<int64_t>> tablet_metadatas;
+    std::vector<string> txn_logs;
+    std::unordered_map<int64_t, std::vector<int64_t>> locked_tablet_metadatas;
```
Comment (wyb) [inline be/src/storage/lake/gc.cpp:38]: map<int64_t, set<int64_t>>?

File: fe/fe-core/src/main/java/com/starrocks/load/ExportJob.java [hunk 7]
```diff
@@ -604,6 +626,33 @@ public Status releaseSnapshotPaths() {
...
+
+            TInternalScanRange internalScanRange = scanRange.getInternal_scan_range();
+            List<TScanRangeLocation> locations = tablet.getLocations();
+            for (TScanRangeLocation location : locations) {
+                TNetworkAddress address = location.getServer();
+                String host = address.getHostname();
+                int port = address.getPort();
+                Backend backend = GlobalStateMgr.getCurrentSystemInfo().getBackendWithBePort(host, port);
+                RpcContext rpcContext = RpcContext.getContext();
+                rpcContext.setReadTimeoutMillis(600000);
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/load/ExportJob.java:644]: default timeout is ok.

File: fe/fe-core/src/main/java/com/starrocks/task/ExportPendingTask.java [hunk 8]
```diff
@@ -112,26 +119,52 @@ private Status makeSnapshots() {
...
+
+    private void lockTabletMetadata(TInternalScanRange internalScanRange, Backend backend) {
+        RpcContext rpcContext = RpcContext.getContext();
+        rpcContext.setReadTimeoutMillis(600000);
+        LakeServiceAsync lakeService = BrpcProxy.getLakeService(backend.getHost(), backend.getBrpcPort());
+        LockTabletMetadataRequest request = new LockTabletMetadataRequest();
+        request.tabletId = internalScanRange.getTablet_id();
+        request.version = Long.parseLong(internalScanRange.getVersion());
+        request.txnId = job.getTxnId();
+        lakeService.lockTabletMetadata(request, new EmptyRpcCallback<>());
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/task/ExportPendingTask.java:143]: Should check the result.

**Comment by TszKitLo40** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 1 / 69 (01.45%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/task/ExportPendingTask.java|0|37|00.00%|[123, 124, 126, 128, 129, 131, 132, 133, 137, 138, 139, 140, 141, 142, 143, 146, 147, 151, 154, 158, 159, 161, 162, 163, 164, 165, 166, 167, 169, 170, 171, 172, 173, 174, 175, 176, 178]|
|:large_blue_circle:|com/starrocks/load/ExportJob.java|1|32|03.12%|[224, 225, 591, 594, 596, 598, 629, 630, 631, 632, 635, 636, 637, 638, 639, 640, 641, 642, 643, 645, 646, 647, 648, 649, 650, 651, 652, 677, 717, 871, 875]|

**Comment by TszKitLo40** [comment]
> #### [FE PR Coverage Check]
> 😞 **fail** : 1 / 69 (01.45%)
> 
> #### file detail
> path	covered_line	new_line	coverage	not_covered_line_detail
> 🔵	com/starrocks/task/ExportPendingTask.java	0	37	00.00%	[123, 124, 126, 128, 129, 131, 132, 133, 137, 138, 139, 140, 141, 142, 143, 146, 147, 151, 154, 158, 159, 161, 162, 163, 164, 165, 166, 167, 169, 170, 171, 172, 173, 174, 175, 176, 178]
> 🔵	com/starrocks/load/ExportJob.java	1	32	03.12%	[224, 225, 591, 594, 596, 598, 629, 630, 631, 632, 635, 636, 637, 638, 639, 640, 641, 642, 643, 645, 646, 647, 648, 649, 650, 651, 652, 677, 717, 871, 875]

It seems that these code is not covered before. @wyb

File: be/src/service/service_be/lake_service.cpp [hunk 9]
```diff
@@ -353,4 +353,74 @@ void LakeServiceImpl::get_tablet_stats(::google::protobuf::RpcController* contro
...
+
+    auto tablet = _env->lake_tablet_manager()->get_tablet(request->tablet_id());
+    if (!tablet.ok()) {
+        LOG(ERROR) << "Fail to get tablet " << request->tablet_id() << ": " << tablet.status();
+        return;
+    }
+    auto st = tablet->delete_txn_log(request->txn_id());
+    if (!st.ok()) {
+        LOG(ERROR) << "Failed to unlock tablet metadata tablet id " << request->tablet_id() << " " << request->txn_id();
+        cntl->SetFailed("Faled to unlock tablet metadata");
```
Comment (wyb) [inline be/src/service/service_be/lake_service.cpp:422]:
```suggestion
        LOG(ERROR) << "Failed to unlock tablet metadata. tablet id: " << request->tablet_id() << ", txn id: " << request->txn_id();
        cntl->SetFailed("Failed to unlock tablet metadata");
```

File: be/src/service/service_be/lake_service.cpp [hunk 10]
```diff
@@ -353,4 +353,74 @@ void LakeServiceImpl::get_tablet_stats(::google::protobuf::RpcController* contro
...
+    }
+    auto txn_log = std::make_shared<starrocks::lake::TxnLog>();
+    txn_log->set_txn_id(request->txn_id());
+    txn_log->set_tablet_id(request->tablet_id());
+    auto op_metadata_lock = txn_log->mutable_op_tablet_metadata_lock();
+    op_metadata_lock->set_version(request->version());
+    auto st = tablet->put_txn_log(std::move(txn_log));
+    if (!st.ok()) {
+        LOG(ERROR) << "Failed to lock tablet metadata tablet id " << request->tablet_id() << " " << request->txn_id();
+        cntl->SetFailed("Faled to lock tablet metadata");
```
Comment (wyb) [inline be/src/service/service_be/lake_service.cpp:388]:
```suggestion
        LOG(ERROR) << "Failed to lock tablet metadata. tablet id: " << request->tablet_id() << ", txn id: " << request->txn_id();
        cntl->SetFailed("Failed to lock tablet metadata");
```

File: be/src/storage/lake/gc.cpp [hunk 11]
```diff
@@ -60,6 +83,12 @@ Status metadata_gc(std::string_view root_location, TabletManager* tablet_mgr) {
         // Keep the latest 10 versions.
```
Comment (wyb) [inline be/src/storage/lake/gc.cpp:83]: Update the comment.

**Review by wyb — DISMISSED** [review]

File: be/src/service/service_be/lake_service.cpp [hunk 12]
```diff
@@ -353,4 +353,79 @@ void LakeServiceImpl::get_tablet_stats(::google::protobuf::RpcController* contro
...
+        LOG(ERROR) << "Fail to lock tablet metadata, tablet id: " << request->tablet_id()
+                   << ", version: " << request->version();
+        cntl->SetFailed("Fail to lock tablet metadata");
+        return;
+    }
+
+    auto tablet_meta = tablet->get_metadata(request->version());
+    // If metadata has been deleted, the request should fail.
+    if (!tablet_meta.ok()) {
+        LOG(ERROR) << "Tablet metadata of tablet " << request->tablet_id() << " has been deleted";
```
Comment (wyb) [inline be/src/service/service_be/lake_service.cpp:393]: Add version in log.

File: be/src/storage/lake/filenames.h [hunk 13]
```diff
@@ -66,4 +75,18 @@ inline std::pair<int64_t, int64_t> parse_txn_log_filename(std::string_view file_
     return {tablet_id, txn_id};
 }
 
+// Return value: <tablet id, txn id>
```
Comment (wyb) [inline be/src/storage/lake/filenames.h:78]:
```suggestion
// Return value: <tablet id, version, expire time >
```

File: be/src/storage/lake/tablet_manager.cpp [hunk 14]
```diff
@@ -657,6 +662,22 @@ Status TabletManager::publish_log_version(int64_t tablet_id, int64_t txn_id, int
     }
 }
 
+Status TabletManager::put_tablet_metadata_lock(int64_t tablet_id, int64_t version, int64_t expire_time) {
+    auto options = WritableFileOptions{.sync_on_close = true, .mode = FileSystem::CREATE_OR_OPEN_WITH_TRUNCATE};
+    auto tablet_metadata_lock_path = tablet_metadata_lock_location(tablet_id, version, expire_time);
+    ASSIGN_OR_RETURN(auto wf, fs::new_writable_file(options, tablet_metadata_lock_path));
+    RETURN_IF_ERROR(wf->append("payload"));
```
Comment (wyb) [inline be/src/storage/lake/tablet_manager.cpp:669]: How about writing an empty proto for future compatibility?

File: be/src/storage/lake/gc.cpp [hunk 15]
```diff
@@ -58,8 +70,15 @@ Status metadata_gc(std::string_view root_location, TabletManager* tablet_mgr) {
         }
         // TODO: batch delete
         // Keep the latest 10 versions.
+        // If the tablet metadata is locked, the correspoding version will be kept.
         std::sort(versions.begin(), versions.end());
         for (size_t i = 0, sz = versions.size() - max_versions; i < sz; i++) {
+            if (locked_tablet_metadatas.count(tablet_id)) {
+                auto locked_tablet_metadata = locked_tablet_metadatas[tablet_id];
```
Comment (sduzh) [inline be/src/storage/lake/gc.cpp:77]:
```suggestion
                const auto& locked_tablet_metadata = locked_tablet_metadatas[tablet_id];
```

**Review by sduzh — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 0 / 67 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/load/ExportJob.java|0|33|00.00%|[583, 586, 588, 590, 621, 622, 623, 624, 627, 628, 629, 630, 631, 632, 633, 634, 635, 638, 639, 640, 641, 642, 643, 644, 645, 646, 647, 648, 649, 650, 675, 715, 869]|
|:large_blue_circle:|com/starrocks/task/ExportPendingTask.java|0|34|00.00%|[121, 122, 124, 126, 127, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 149, 150, 152, 153, 154, 155, 156, 157, 158, 160, 161, 162, 163, 164, 165, 166, 167, 169, 170]|

**Review by wyb — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by TszKitLo40** [comment]
run starrocks_admit_test

**Pull request merged**